### DISCUSSION TITLE
Cold lambda functions on Vercel are very slow

### DISCUSSION THREAD

**Discussion by kachkaev** [vercel/community#496] [Show & Tell]
👋 folks! I help develop [blockprotocol.org](https://blockprotocol.org), which is an open source project hosted on GitHub: [blockprotocol/blockprotocol](https://github.com/blockprotocol/blockprotocol).

We deploy to Vercel and observe very slow Lambda functions when they are not warmed up. This applies to pages with `getServerSideProps` or `getStaticProps + fallback` as well as to API routes. Even if a Lambda function does not contain any payload, it may take ~10 seconds to execute. This suggests that it’s not the function execution time that is slow but rather its warm-up phase (i.e. something that we as app developers do not control). Subsequent calls to warmed-up Lambdas are very quick, as expected.

Here is some profiling for a deployment in https://github.com/blockprotocol/blockprotocol/pull/268

### Home page – `/`

This page is rendered statically so is always quick. However, a call to `/api/me` takes ~10 seconds for the first person that hits this URL. `401` is expected for a user that is not logged in.

<img width="822" alt="Screenshot 2022-03-15 at 11 47 19" src="https://user-images.githubusercontent.com/608862/158375854-6ce44481-bff1-45e2-881c-6c8a252256e8.png">

[vercel-speed-homepage.har.zip](https://github.com/vercel/community/files/8253148/vercel-speed-homepage.har.zip)

### Profile page – `/@hash`

This page uses `getServerSideProps` – see https://github.com/blockprotocol/blockprotocol/pull/268. The initial render is slow, but a reload (take 2) is quick. Calls to `/api/me` are always quick below because this API route has been warmed up previously.

### Take 1

<img width="824" alt="Screenshot 2022-03-15 at 11 48 44" src="https://user-images.githubusercontent.com/608862/158376344-899ee8d1-8f6a-4015-914f-8d3770ec937e.png">

[vercel-speed-profile-page.har.zip](https://github.com/vercel/community/files/8253162/vercel-speed-profile-page.har.zip)

### Take 2

<img width="822" alt="Screenshot 2022-03-15 at 11 52 37" src="https://user-images.githubusercontent.com/608862/158376772-c1b06778-371e-488c-92a8-9676ad91506d.png">

[vercel-speed-profile-page-take-2.har.zip](https://github.com/vercel/community/files/8253188/vercel-speed-profile-page-take-2.har.zip)

We use `getStaticProps` + `revalidate` in the `main` branch instead of `getServerSideProps` [since recently](https://github.com/blockprotocol/blockprotocol/pull/250/files#diff-1ed99ec657b21c17e18966d43769d1cde4da6cf404eb2d3f665495621f9e4336). It is also slow on first load. I have switched back to `getServerSideProps` to make the demonstration cleaner.

---

The problem with this issue is that it’s quite hard to reproduce or investigate. As soon as someone has opened a page, all subsequent loads are quick for a given deployment. If a page has not been visited for several minutes, it may become slow again and takes the same ~10 seconds to load as for the first time. The longest part of each slow request is TTFB:

<img width="595" alt="Screenshot 2022-03-15 at 12 29 23" src="https://user-images.githubusercontent.com/608862/158377978-742cbee4-a2c2-481e-a52d-6f183247a087.png">

This issue was raised several times, e.g. see:
- https://github.com/vercel/community/discussions/427
- https://github.com/vercel/vercel/discussions/5422
- https://github.com/vercel/vercel/issues/6292

@Pieparker [asked](https://github.com/vercel/community/discussions/427#discussioncomment-2339552) me to report our specific case. Happy to help with the investigation!

In the meantime, we warm-up slow pages via GitHub Actions: https://github.com/blockprotocol/blockprotocol/pull/250

**Comment by Pieparker** [comment]
Hey there. I appreciate you reaching out with this super insightful report!

I've checked out the first HAR file for your index page. Most resources take between 4-800ms. Seems reasonable! I can see the longer wait times for, for example, `/api/me`. Let me dig in with our wider engineering team ([internal link only](https://github.com/vercel/customer-issues/issues/360)) and investigate what's happening. Hang tight!

**Reply by Pieparker** [reply]
Hey @kachkaev, following up again here! If possible, whilst we dig in, it would be worth testing the blank getServerSideProps on a Next.js starter. This will give us a benchmark and prove/disprove other extraneous variables!

**Reply by kachkaev** [reply]
Hi @Pieparker! Are you suggesting to empty the codebase in my experimental PR (https://github.com/blockprotocol/blockprotocol/pull/268) and replace it with Next.js starter? Just trying to understand what you would like me to test.

In the meantime, I created `/site/src/pages/simple-ssr-page.page.tsx` with the following contents:

```js
import { NextPage, GetServerSideProps } from "next";

interface SimpleSsrPageProps {
  message: string;
}

export const getServerSideProps: GetServerSideProps<
  SimpleSsrPageProps
> = async () => {
  return {
    props: { message: "hello world" },
  };
};

const SimpleSsrPage: NextPage<SimpleSsrPageProps> = ({ message }) => {
  return <div>{message}</div>;
};

export default SimpleSsrPage;
```

Here are the benchmarks:

#### Take 1

<img width="781" alt="Screenshot 2022-03-16 at 18 11 40" src="https://user-images.githubusercontent.com/608862/158662206-f443452e-1853-4289-b830-facba09705cf.png">

[vercel-speed-simple-ssr-page.har.zip](https://github.com/vercel/community/files/8268676/vercel-speed-simple-ssr-page.har.zip)

#### Take 2

<img width="777" alt="Screenshot 2022-03-16 at 18 26 05" src="https://user-images.githubusercontent.com/608862/158662232-adf29922-bd70-4b01-ad90-13ee80c47189.png">

[vercel-speed-simple-ssr-page-take-2.har.zip](https://github.com/vercel/community/files/8268679/vercel-speed-simple-ssr-page-take-2.har.zip)

The new page still uses `_document` and `_app` which has some Tailwind and Emotion boilerplate. I don’t believe that they affect Lamda performance, because if they did, we did not observe slow cold API requests (`/api/me`). 🤔

**Reply by Pieparker** [reply]
Oh no, we'd never ask you to strip out production code! Instead, it would be great to set up the most basic of functions – a blank getServerSideProps. I appreciate you sharing all of this with me and I'll feed back once the team have questions, and/or something worth relaying to you 😄!

**Reply by kachkaev** [reply]
👋 @Pieparker what would be the next steps?

**Reply by vilkinsons** [reply]
Hey @Pieparker, just chasing this. If there's anything we can do to help get this resolved, please let me know.

**Reply by Pieparker** [reply]
Still waiting on updates internally. Appreciate your patience!

**Reply by vilkinsons** [reply]
Our website is unusably slow in this regard; we're going to start scoping alternative approaches/providers next week if we're unable to advance this before then.

**Reply by Pieparker** [reply]
Understood! What I can share so far is:
- From our sampled logs, your max duration for a serverless function invocation since 10th March has been 3.4s (with one single exception on the 25th March of 14s)
- This behaviour does not correlate 1:1 with new deployments
- It would appear as though a change was introduced in the deployment with ID `6JJr582umZKudmzgwikDGMnkUEM3` which remediated the long Lambda function invocations
  - Prior to this deployment, all deployments were showing frequent 10/11s invocations

There's nothing we can identify systemically to suggest that there is an underlying problem, with a fraction of a percent's worth of invocations taking >10s (and these are likely either by design/due to sluggish backends/extraneous variables).

I'd be interested to learn what changes were made in the above deployment and whether you're still observing such latent requests.

**Reply by Pieparker** [reply]
Hey there, team! We've had a response from our engineering team. I'll summarise:

- The app comes in @ around 15MB (Lambda) size, so it takes quite a while to evaluate that much JavaScript. **The remediation here would be to reduce the application size.**
- In testing, we did increase the amount of RAM the function has from 1GB to 3GB and this succeeded in reducing the cold boots by 50%, reinforcing the fact that this is a CPU limitation on code evaluation
- In testing, we could observe 5s `init` (cold starts) which were reduced to 1-2s after increasing memory
  - 2.5s was spent in user code (setting up database for the first time, connecting to Mongo etc)
  - Remaining time (~2s) was spent in transit
- Next.js' Open Telemetry work should make cases like this much more apparent (it'd let you know when all the time is being spent in code init phase, etc)

Thanks ever so much for your patience, and I hope that this helps!

**Reply by kachkaev** [reply]
Thanks for looking into our case @Pieparker! I added `@next/bundle-analyzer` to our project via https://github.com/blockprotocol/blockprotocol/pull/286 but could not find the source of 15MB of JS you referred to. Do Lambdas also contain code outside `.next` folder (e.g. some `node_modules` that are not bundled so don't appear in the stats)?

I checked deployment `6JJr582umZKudmzgwikDGMnkUEM3` you mentioned – it resolved to https://github.com/blockprotocol/blockprotocol/pull/222. This change does not add any dependencies to the project (a block is not a part of the Next.js app). Nearby PRs from around 18 February don’t look suspicious either.

Unfortunately, our cold Lambda calls are still slow, which you can observe here: https://github.com/blockprotocol/blockprotocol/actions/workflows/warm-up-vercel.yml?query=event%3Adeployment_status

**[Example](https://github.com/blockprotocol/blockprotocol/runs/6091952536)** (for https://github.com/blockprotocol/blockprotocol/pull/286):

```
  https://blockprotocol-3tcqaq7v1-hashintel.vercel.app/@hash
    Status code: 200
    Time: 17.981105
  
  https://blockprotocol-3tcqaq7v1-hashintel.vercel.app/api/me
    Status code: 401
    Time: 0.130181
  
  https://blockprotocol-3tcqaq7v1-hashintel.vercel.app/types/DUMMY_ENTITY_ID
    Status code: 404
    Time: 7.998259
```

[GitHub Workflow source](https://github.com/blockprotocol/blockprotocol/blob/9c6ab2d2d5b76b981bca1a86cbebcd0a2e516e91/.github/workflows/warm-up-vercel.yml)

**Reply by Pieparker** [reply]
Hey again!

You can view the bundle size at any time via the function log section of the UI: https://vercel.com/[team-name]/[project-name] -> `View Function Logs` -> [select a function]. You'll see that `/api/me` is just shy of 15MB.

[Multiple functions are bundled together](https://vercel.com/docs/concepts/functions/serverless-functions#bundling-serverless-functions), which will likely explain the size disparity. As a test, you can force them to be bundled separately by specifying unique properties via [functions config](https://vercel.com/docs/project-configuration#project-configuration/functions). This is where you'd also be able to specify [up to 3008MB of memory](https://vercel.com/docs/concepts/limits/overview#serverless-function-memory) on a paid plan. The `init` time (or the "cold start") is truly only 1 to 2 seconds. Therefore, by reducing the bundle size and providing more memory, we will be able to take seconds off of the `evaluation` time.

**Reply by kachkaev** [reply]
👋 @Pieparker! Thanks for the tips!

I looked at functions stats for one of the recent deployments and discovered this:

| Function path | Memory Size | Size | Maximum Execution Duration |
|:-|-:|-:|-:|
`[shortname]/[[...profileTabs]]` | 1024 MB | 16.15 MB | 1 minute
`[shortname]/blocks/[blockSlug]` | 1024 MB | 16.15 MB | 1 minute
`api/blocks` | 1024 MB | 14.27 MB | 1 minute
`api/completeSignup` | 1024 MB | 14.27 MB | 1 minute
`api/fetchEmbedCode` | 1024 MB | 14.27 MB | 1 minute
`api/isShortnameTaken` | 1024 MB | 14.27 MB | 1 minute
`api/loginWithLoginCode` | 1024 MB | 14.27 MB | 1 minute
`api/logout` | 1024 MB | 14.27 MB | 1 minute
`api/me` | 1024 MB | 14.27 MB | 1 minute
`api/me/generateApiKey` | 1024 MB | 14.27 MB | 1 minute
`api/sendLoginCode` | 1024 MB | 14.27 MB | 1 minute
`api/signup` | 1024 MB | 14.27 MB | 1 minute
`api/types/[id]/update` | 1024 MB | 14.27 MB | 1 minute
`api/types/create` | 1024 MB | 14.27 MB | 1 minute
`api/uploadUserAvatar` | 1024 MB | 14.27 MB | 1 minute
`api/users/[shortname]` | 1024 MB | 14.27 MB | 1 minute
`api/users/[shortname]/blocks` | 1024 MB | 14.27 MB | 1 minute
`api/users/[shortname]/types` | 1024 MB | 14.27 MB | 1 minute
`api/users/[shortname]/types/[title]` | 1024 MB | 14.27 MB | 1 minute
`api/verifyEmail` | 1024 MB | 14.27 MB | 1 minute
`docs/[[...docsSlug]]` | 1024 MB | 16.15 MB | 1 minute
`types/[entityTypeId]` | 1024 MB | 11.90 MB | 1 minute


Because there are a lot of repeating size values, it indeed looks as if Vercel is bundling the functions together. That’s strange because it looks like that’s what could be causing slow warm up times.

I tried configuring function params in https://github.com/blockprotocol/blockprotocol/pull/290 by creating `site/vercel.json`. This was not very successful. When I [add](https://github.com/blockprotocol/blockprotocol/pull/290/commits/1a685effd84bdfe2668b91e8951a95b08543009b#diff-5acaaa180fde5dcbd944c663dc9fe9d54042a8b94fea18d593fe412f0b8e6caa)

```js
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "functions": {
    "pages/[shortname]/[[...profileTabs]].page.tsx": {
      "memory": 3008,
      "maxDuration": 30
    },
    "pages/api/me.api.ts": {
      "memory": 2880,
      "maxDuration": 30
    }
  }
}
```

nothing changes, but when I [go for](https://github.com/blockprotocol/blockprotocol/pull/290/commits/404f0b6741aa11679065ac688b51ea0ab6e927a9)

```js
{
  "$schema": "https://openapi.vercel.sh/vercel.json",
  "functions": {
    "[shortname]/[[...profileTabs]].js": {
      "memory": 3008,
      "maxDuration": 30
    },
    "api/me.js": {
      "memory": 2880,
      "maxDuration": 30
    }
}
```

instead, deployment fails immediately:

```
Error: The pattern "[shortname]/[[...profileTabs]].js" defined in `functions` doesn't match any Serverless Functions. Learn More: https://vercel.link/unmatched-function-pattern
```


I also see this in the first case:

```
WARNING: Unable to find source file for page api/blocks.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
WARNING: Unable to find source file for page api/completeSignup.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
WARNING: Unable to find source file for page api/fetchEmbedCode.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
WARNING: Unable to find source file for page api/isShortnameTaken.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
WARNING: Unable to find source file for page api/loginWithLoginCode.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
WARNING: Unable to find source file for page api/logout.js with extensions: js, jsx, ts, tsx, this can cause functions config from `vercel.json` to not be applied
...
```

Same as here: https://github.com/vercel/next.js/discussions/23999

What’s the right location of `vercel.json` and what are the right function paths? The app is in `site` workspace and the source is in `site/src/pages/(**/*.page.tsx|api/**/*.api.ts)`.

Happy to continue the experiment with your help!

**Reply by Pieparker** [reply]
Hey! `vercel.json` should reside in the `/` directory typically. 

You should map to the page file, e.g. `pages/api/hello.ts`, where applicable. Perhaps there's a nuance with page files and this config. @ijjk would you be able to confirm what options, if any, we have for preventing bundling for this use case above?

**Reply by ijjk** [reply]
Hi, it seems we aren't detecting the functions config correctly due to the custom `pageExtensions` being used which we can look into handling. 

Taking a quick look at the shared repository, the cold boot times are most likely related to loading large dependencies e.g. these icon libraries https://github.com/blockprotocol/blockprotocol/blob/24f2b21ca8e228337f57464caf6eb79e63441750/site/package.json#L26-L30 specifically the `@material/material-icons` library can cause thousands of imports which need to be parsed when the stub file is loaded instead of the exact icon.

**Reply by vilkinsons** [reply]
@ijjk Are there any plans for a fix re: this? Thanks!

**Reply by ijjk** [reply]
@nonparibus the fix for the above issue would be to ensure you are only importing necessary files e.g. instead of the stub file `@mui/icons-material` you should import the specific icon `@mui/icons-material/ChevronLeft`. This prevents the need to process a massive number of un-necessary imports

**Reply by kachkaev** [reply]
Hmmm what stops Next.js from tree-shaking `@mui/icons-material`? I also wonder if [`outputStandalone`](https://nextjs.org/docs/advanced-features/output-file-tracing#automatically-copying-traced-files-experimental) can be a solution here.

**UPD**: I tried `outputStandalone` as well as avoiding imports from Icon library roots in https://github.com/blockprotocol/blockprotocol/pull/393. However, this did not change anything. `/api/*` lambdas on Vercel are still 10+MB despite rather modest bundle sizes in `yarn build` output:

```
Page                                                                Size     First Load JS
┌ ● /                                                               191 B           295 kB
├   /_app                                                           0 B             295 kB
...
├ ○ /404                                                            194 B           295 kB
├ λ /api/blocks                                                     0 B             295 kB
├ λ /api/complete-signup                                            0 B             295 kB
├ λ /api/is-shortname-taken                                         0 B             295 kB
├ λ /api/login-with-login-code                                      0 B             295 kB
├ λ /api/logout                                                     0 B             295 kB
├ λ /api/me                                                         0 B             295 kB
├ λ /api/me/api-keys                                                0 B             295 kB
...
```

Just as before, several lambda functions report identical sizes in Vercel UI, which suggests that they have been grouped together by the platform. Not sure how to split them; see what I’ve tried in https://github.com/orgs/vercel/discussions/496#discussioncomment-2615949. Any help will be appreciated!

**Reply by ijjk** [reply]
@kachkaev you can set a `NEXT_DEBUG_FUNCTION_SIZE=1` env variable on your project and it will provide some additional information, `outputStandalone` just creates a folder using the traces to allow deploying without `node_modules` for self-hosting. Vercel uses the same traces but also bundles. 

Note: this bundling shouldn't have much affect on cold boot times unless the related code is needed for the specific page being requested as the code is lazily loaded for each page as it is needed, this actually helps keep the functions warm longer. 

You should also seem similar slow load times requiring the necessary bundles locally although may not be as noticeable if you are on a high end laptop compared to lambda e.g. 
```js
const timeLabel = 'Loaded page bundle'
console.time(timeLabel)
require('./.next/server/pages/_app.js')
require('./.next/server/pages/_document.js')
require('./.next/server/pages/page-path.js')
console.timeEnd(timeLabel)
```

**Reply by kachkaev** [reply]
👋 @ijjk! I created `site/timing.cjs` locally with this content:

```js
const timeLabel = "Loaded page bundle";

console.time(timeLabel);

require("dotenv-flow").config();
// require("./.next/server/pages/_app.js");
// require("./.next/server/pages/_document.js");
require("./.next/server/pages/api/me.js");

console.timeEnd(timeLabel);
```

Running `cd site; yarn build; node ./timing.cjs` reports about 300ms and uncommenting `_app` and `_document` only doubles this number. This is nowhere near ≈10 seconds on Vercel.

Thanks for the tip about `NEXT_DEBUG_FUNCTION_SIZE=1`! I set this variable on Vercel and was able to see how the lambdas were grouped.

<details>
<summary>Vercel build output</summary>
<pre>
...
Serverless function size info
Serverless Function's pages: [shortname]/[[...profile-tabs]].js, [shortname]/blocks/[block-slug].js, docs/[[...docs-slug]].js, hub.js, index.js
Large Dependencies          Uncompressed size  Compressed size
node_modules/@mui/material            1.86 MB           568 kB
node_modules/react-dom/cjs            1.02 MB           244 kB
node_modules/@mui/base                 406 kB           119 kB
site/.next/server                      525 kB           108 kB
node_modules/acorn/dist                425 kB           103 kB
All dependencies                      22.3 MB          6.43 MB
Serverless Function's page: types/[entity-type-id].js
Large Dependencies          Uncompressed size  Compressed size
site/public/blocks                    12.6 MB           3.7 MB
node_modules/@mui/material            1.86 MB           568 kB
node_modules/react-dom/cjs            1.02 MB           244 kB
node_modules/mongodb/lib               770 kB           188 kB
node_modules/@mui/base                 406 kB           119 kB
All dependencies                      20.7 MB          5.86 MB
Serverless Function's pages: api/blocks.js, api/complete-signup.js, api/is-shortname-taken.js, api/login-with-login-code.js, api/logout.js, api/me.js, api/me/api-keys.js, api/me/generate-api-key.js, api/rewrites/sandboxed-block-demo.js, api/send-login-code.js, api/signup.js, api/types/[id].js, api/types/[id]/update.js, api/types/create.js, api/upload-user-avatar.js, api/users/[shortname].js, api/users/[shortname]/blocks.js, api/users/[shortname]/types.js, api/users/[shortname]/types/[title].js, api/verify-email.js
Large Dependencies                Uncompressed size  Compressed size
site/public/blocks                          12.6 MB           3.7 MB
node_modules/rword/words                    5.81 MB          1.41 MB
node_modules/@mui/material                  1.86 MB           568 kB
node_modules/react-dom/cjs                  1.02 MB           244 kB
node_modules/mongodb/lib                     770 kB           188 kB
site/.next/server                            568 kB           119 kB
node_modules/@mui/base                       406 kB           119 kB
node_modules/@aws-sdk/client-s3              946 kB           110 kB
node_modules/nodemailer/lib                  438 kB           101 kB
node_modules/lodash/lodash.js                544 kB          96.4 kB
node_modules/@aws-sdk/client-ses             620 kB          73.8 kB
All dependencies                            32.8 MB          8.55 MB
Created all serverless functions in: 7.179s
Collected static files (public/, static/, .next/static): 39.647ms
Generated build outputs:
 - Static files: 295
 - Serverless Functions: 81
 - Edge Functions: 0
Serverless regions: Washington, D.C., USA
Deployed outputs in 5s
...
</pre>
</details>

This is an interesting bit:

```
Serverless Function's pages: api/blocks.js, api/complete-signup.js, api/is-shortname-taken.js,
api/login-with-login-code.js, api/logout.js, api/me.js, api/me/api-keys.js, api/me/generate-api-key.js,
api/rewrites/sandboxed-block-demo.js, api/send-login-code.js, api/signup.js, api/types/[id].js,
api/types/[id]/update.js, api/types/create.js, api/upload-user-avatar.js, api/users/[shortname].js,
api/users/[shortname]/blocks.js, api/users/[shortname]/types.js, api/users/[shortname]/types/[title].js,
api/verify-email.js
```

So most API routes are grouped together, which creates a pretty sizeable container with quite a few dependencies. Before Node is even started, the container must be downloaded and spawned by AWS, and this is probably where ≈10 seconds are spent. At least this explains this large difference between cold and warm starts to me.

I _guess_ this ‘overgrouping’ may be related to our usage of [`models`](https://github.com/blockprotocol/blockprotocol/tree/9645db355f270468c1473ca41f714ba522e13b86/site/src/lib/api/model). We’ve organised some logic as classes that cannot be tree-shaken. Calling a single method from a model pulls all dependencies that are needed for all methods. I’ll speak to my team mates about refactoring models – curious if this can help.

Although the problem is still there, I’m glad that it is now transparent thanks to `NEXT_DEBUG_FUNCTION_SIZE=1`. If you or your colleagues know how to force-split the groups, happy to try that too! My previous attempts [were not successful](https://github.com/orgs/vercel/discussions/496#discussioncomment-2615949) (not sure what exactly to define).

**Reply by widavies** [reply]
@kachkaev Did you end up finding a fix for this?

**Reply by kachkaev** [reply]
The only fix is to reduce lambda size, I am afraid. This may imply moving some business logic away from Vercel (or any other Lambda-based infra). As a workaround, you can try poking your API endpoints time and again (https://github.com/blockprotocol/blockprotocol/pull/250) to reduce a chance of your app users hitting cold lambdas.
